ArduPlane/soaring.cpp

```diff
@@ -66,156 +66,118 @@ void Plane::update_soaring() {
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooFar, mode_cruise.name());
-                    set_mode(mode_cruise, ModeReason::SOARING_DRIFT_EXCEEDED);
-                    break;
-            } // switch loiterStatus
-            cruise_state = cruise;
-            set_target_altitude_current();
-            break;
-            } // case Cruise
+        // Determine appropriate mode.
+        Mode* exit_mode = previous_mode;
 
-        case Mode::Number::AUTO: {
-            //Get the lat/lon of next Nav waypoint after this one:
-            AP_Mission::Mission_Command current_nav_cmd = mission.get_current_nav_cmd();;
-            const bool headingLinedupToWP = plane.mode_loiter.isHeadingLinedUp(next_WP_loc, current_nav_cmd.content.location);
-            if (!headingLinedupToWP && loiterStatus!=SoaringController::LoiterStatus::ALT_TOO_LOW && timer<20e3) {
+        if (loiterStatus == SoaringController::LoiterStatus::ALT_TOO_LOW && 
+           ((previous_mode == &mode_cruise) || (previous_mode == &mode_fbwb))) {
+            exit_mode = &mode_rtl;
+        }
+
+        // Print message and set mode.
+        switch (loiterStatus) {
+            case SoaringController::LoiterStatus::ALT_TOO_HIGH:
+                gcs().send_text(MAV_SEVERITY_INFO, "Soaring: Too high, restoring %s", exit_mode->name());
+                set_mode(*exit_mode, ModeReason::SOARING_ALT_TOO_HIGH);
                 break;
-            }
-            switch (loiterStatus) {
-                case SoaringController::LoiterStatus::ALT_TOO_HIGH:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooHigh, mode_auto.name());
-                    set_mode(mode_auto, ModeReason::SOARING_ALT_TOO_HIGH);
-                    break;
-                case SoaringController::LoiterStatus::ALT_TOO_LOW:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooLow, mode_auto.name());
-                    set_mode(mode_auto, ModeReason::SOARING_ALT_TOO_LOW);
-                    break;
-                default:
-                case SoaringController::LoiterStatus::THERMAL_WEAK:
-                        gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooWeak, mode_auto.name());
-                        set_mode(mode_auto, ModeReason::SOARING_THERMAL_ESTIMATE_DETERIORATED);
-                    break;
-                case SoaringController::LoiterStatus::DRIFT_EXCEEDED:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooFar, mode_auto.name());
-                    set_mode(mode_auto, ModeReason::SOARING_DRIFT_EXCEEDED);
-                    break;
-                } // switch loiterStatus
-            break;
-            } // case AUTO
-        default: // all other modes
-            break;
-        } // switch previous_mode
+            case SoaringController::LoiterStatus::ALT_TOO_LOW:
+                gcs().send_text(MAV_SEVERITY_INFO, "Soaring: Too low, restoring %s", exit_mode->name());
```

Comment:
Might save a few bytes by creating a "soaring_restore_mode":

```
soaring_restore_mode("Too low", ModeReason::SOARING_ALT_TOO_LOW, *exit_mode);
```

```
void Plane::soaring_restore_mode(const char *reason, ModeReason modereason, Mode &exit_mode)
{
     gcs().send_text(MAV_SEVERITY_INFO, "Soaring: %s, restoring %s", exit_mode.name());
     set_mode(exit_mode, modereason);
}

```

A lookup table from SoaringController::LoiterStatus -> const char*description might make reuse of the strings possible, too.


Not a blocker.

Reply:
I was thinking this same thing

Reply:
Great, added this helper method.